### PR TITLE
allow a new upgrade version without DB changes to be inserted

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #6760

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

unit tests (so far)

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseUpgradeCheckerTest.java
Comment (DaanHoogland) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseUpgradeCheckerTest.java:170]:
```suggestion
        assertEquals("We should have 2 upgrade steps", 2, upgrades.length);
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@rohityadavcloud @nvazquez @borisstoyanov can you review? I think this will be hard to test untill we reach a point in time where it is useful, but it has unit tests (added). Maybe give extra attention to the assertions in those during review!?

File: engine/schema/pom.xml
Comment (rohityadavcloud) [inline engine/schema/pom.xml:63]: should we specify some version here? @DaanHoogland

File: engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java
Comment (rohityadavcloud) [inline engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java:134]: Should we just remove the last digit (from a.b.c.d; remove .d) in the upgrade paths map and the logic that obtains current ACS version (to report a.b.c only to DB upgrade path processors?)

**Review by rohityadavcloud — APPROVED** [review]
Needs testing but LGTM if there are no regression tests

File: engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java
Comment (DaanHoogland) [reply engine/schema/src/main/java/com/cloud/upgrade/DatabaseVersionHierarchy.java:134]: unfortunately a lot of four digit versions are already in the DBs around the world. In addition to that, we might want to add a DB patch in a security release at some pioint in which case we want to take the fourth digit into account.

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java
Comment (weizhouapache) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:35]: @DaanHoogland 
what's the difference between this and `Assert.assertEquals` ?

Comment (DaanHoogland) [reply engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:35]: this is junit5 as opposed to junit4

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]

**Comment by weizhouapache** [comment]
@blueorangutan package

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java
Comment (DaanHoogland) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:26]:
```suggestion
import com.cloud.upgrade.dao.Upgrade41720to41800;
```

File: engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java
Comment (DaanHoogland) [inline engine/schema/src/test/java/com/cloud/upgrade/DatabaseVersionHierarchyTest.java:93]:
```suggestion
                .next("4.17.2.0", new Upgrade41710to41800());
```

**Review by DaanHoogland — COMMENTED** [review]
semantic conflict :(

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Pull request merged**
